servicetalk-log4j2-mdc/src/main/java/io/servicetalk/log4j2/mdc/MdcCapturedContextProvider.java

```diff
@@ -0,0 +1,109 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.log4j2.mdc;
+
+import io.servicetalk.concurrent.api.CapturedContext;
+import io.servicetalk.concurrent.api.CapturedContextProvider;
+import io.servicetalk.concurrent.api.Scope;
+import io.servicetalk.context.api.ContextMap;
+
+import org.apache.logging.log4j.ThreadContext;
+import org.apache.logging.log4j.spi.ReadOnlyThreadContextMap;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A {@link CapturedContextProvider} implementation to correctly propagate MDC context.
+ * Note: this class should not be used directly: it is only intended to be service loaded by ServiceTalk.
+ */
+public final class MdcCapturedContextProvider implements CapturedContextProvider {
+
+    private final boolean enabled;
+
+    /**
+     * Create a new {@link CapturedContextProvider} for MDC.
+     * Note: this class should not be used directly: it is only intended to be service loaded by ServiceTalk.
+     */
+    public MdcCapturedContextProvider() {
+        enabled = shouldEnableMdcCapture();
+    }
+
+    @Override
+    public CapturedContext captureContext(CapturedContext underlying) {
+        return enabled ? new MdcCapturedContext(underlying, getCurrent()) : underlying;
+    }
+
+    @Override
+    public CapturedContext captureContextCopy(CapturedContext underlying) {
+        return enabled ? new MdcCapturedContext(underlying, getCurrentCopy()) : underlying;
+    }
+
+    private static final class MdcCapturedContext implements CapturedContext {
+
+        private final CapturedContext delegate;
+        private final Map<String, String> storage;
+
+        MdcCapturedContext(CapturedContext delegate, Map<String, String> storage) {
+            this.delegate = delegate;
+            this.storage = storage;
+        }
+
+        @Override
+        public ContextMap captured() {
+            return delegate.captured();
+        }
+
+        @Override
+        public Scope attachContext() {
+            Map<String, String> old = getCurrent();
+            setCurrent(storage);
+            Scope delegateScope = delegate.attachContext();
+            return () -> {
+                delegateScope.close();
+                setCurrent(old);
+            };
+        }
+    }
+
+    private static Map<String, String> getCurrent() {
+        return DefaultServiceTalkThreadContextMap.CONTEXT_STORAGE.get();
+    }
+
+    private static Map<String, String> getCurrentCopy() {
+        Map<String, String> current = DefaultServiceTalkThreadContextMap.CONTEXT_STORAGE.get();
+        Map<String, String> result = new ConcurrentHashMap<>(Math.max(current.size(), 4));
```

Comment:
I see that the `DefaultServiceTalkThreadContextMap.CONTEXT_STORAGE` just deals with a regular Map, but we alaways set a concurrent hash map. Should the type be changed so in a type safe manner we always expect something thread safe (i.e. like the ConcurrentMap)? I was thinking about it since setCurrent in theory expects any map. I see that it is only called from inside the class and I would assume the getCurrent always returns a thread safe one, so if you think it's not needed I can live with it too. 

I was thinking about it too because the getCurrentCopy also makes sure it returns a concurrent one.